mongodb/transaction_test.go

```diff
@@ -0,0 +1,255 @@
+// The transactionCtx is derived from the MongoConnection conn
+// The interleave boolean value allows injection a mongo call outside of the main transaction in transactionCtx, to
+// simulate a parallel transaction that interferes with the main transaction. Without retrying the main transaction,
+// this will lead to a Transaction Error on commit
+func exampleTransactionFunc(conn *mongoDriver.MongoConnection, interleave bool) mongoDriver.TransactionFunc {
+	return func(transactionCtx context.Context) (interface{}, error) {
+		var obj1, obj2 simpleObject
+		err := conn.Collection(collection1).FindOne(transactionCtx, bson.M{"_id": 1}, &obj1)
+		if err != nil {
+			return nil, fmt.Errorf("could not find object in collection (%s): %w", collection1, err)
+		}
+
+		switch obj1.State {
+		case "first":
+			obj1.State = "second"
+		case "second":
+			obj1.State = "third"
+
+			obj2 = simpleObject{ID: 1, State: "final"}
+			_, err = conn.Collection(collection2).Upsert(transactionCtx, bson.M{"_id": obj2.ID}, bson.M{"$set": obj2})
+			if err != nil {
+				return nil, fmt.Errorf("could not upsert object in collection (%s): %w", collection2, err)
+			}
+		default:
+			return nil, badObjectState
+		}
+
+		if interleave {
+			_, err = conn.Collection(collection1).Update(context.Background(), bson.M{"_id": 1}, bson.M{"$set": bson.M{"state": "second"}})
+			if err != nil {
+				return nil, fmt.Errorf("interleave write failed in collection (%s): %w", collection1, err)
+			}
+		}
+
+		_, err = conn.Collection(collection1).Update(transactionCtx, bson.M{"_id": 1}, bson.M{"$set": obj1})
+		if err != nil {
+			return nil, fmt.Errorf("could not write object in collection (%s): %w", collection1, err)
+		}
+
+		return struct{ o1, o2 simpleObject }{o1: obj1, o2: obj2}, nil
+	}
+}
+
+func TestTransaction(t *testing.T) {
+	ctx := context.Background()
+	conn, cleanup := setupMongoConnection(t)
+	defer cleanup(ctx)
+
+	Convey("Given a mongo server cluster", t, func() {
+		Convey("setup with a test simpleObject in 'first' State in test-collection-1", func() {
+			setupTest(t, conn, collection1, simpleObject{ID: 1, State: "first"})
+
+			Convey("when the example transaction is run with neither retries nor an interleaved outside transaction", func() {
+				r, e := conn.RunTransaction(ctx, false, exampleTransactionFunc(conn, false))
+				Convey("the transaction completes successfully, amd the results are as expected", func() {
+					So(e, ShouldBeNil)
+					res, ok := r.(struct{ o1, o2 simpleObject })
+					So(ok, ShouldBeTrue)
+					So(res.o1, ShouldResemble, simpleObject{ID: 1, State: "second"})
+					So(res.o2, ShouldResemble, simpleObject{})
+				})
+			})
+
+			Convey("when the example transaction is run with retries (but without an interleaved outside transaction)", func() {
+				r, e := conn.RunTransaction(ctx, true, exampleTransactionFunc(conn, false))
+				Convey("again the transaction completes successfully, and the results are the same, as expected", func() {
```

Comment:
Again? I think we should describe them independently

```suggestion
				Convey("the transaction completes successfully, and the results are as expected", func() {
```